cmd/relay-gossip/node/relaynode.go

```diff
@@ -0,0 +1,201 @@
+package node
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"time"
+
+	"github.com/drand/drand/cmd/relay-gossip/lp2p"
+	"github.com/drand/drand/protobuf/drand"
+	"github.com/gogo/protobuf/proto"
+	bds "github.com/ipfs/go-ds-badger2"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/prometheus/common/log"
+	"golang.org/x/xerrors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// GossipRelayConfig configures a gossip relay node.
+type GossipRelayConfig struct {
+	Network         string
+	PeerWith        []string
+	Addr            string
+	DataDir         string
+	IdentityPath    string
+	CertPath        string
+	Insecure        bool
+	DrandPublicGRPC string
+}
+
+// GossipRelayNode is a gossip relay runtime.
+type GossipRelayNode struct {
+	bootstrap []ma.Multiaddr
+	ds        *bds.Datastore
+	priv      crypto.PrivKey
+	h         host.Host
+	ps        *pubsub.PubSub
+	t         *pubsub.Topic
+	opts      []grpc.DialOption
+	addr      []ma.Multiaddr
+	done      chan struct{}
+}
+
+// NewGossipRelayNode starts a new gossip relay node.
+func NewGossipRelayNode(cfg *GossipRelayConfig) (*GossipRelayNode, error) {
+	bootstrap, err := ParseMultiaddrSlice(cfg.PeerWith)
+	if err != nil {
+		return nil, xerrors.Errorf("parsing peer-with: %w", err)
+	}
+
+	ds, err := bds.NewDatastore(cfg.DataDir, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("opening datastore: %w", err)
+	}
+
+	priv, err := lp2p.LoadOrCreatePrivKey(cfg.IdentityPath)
+	if err != nil {
+		return nil, xerrors.Errorf("loading p2p key: %w", err)
+	}
+
+	h, ps, err := lp2p.ConstructHost(ds, priv, cfg.Addr, bootstrap)
+	if err != nil {
+		return nil, xerrors.Errorf("constructing host: %w", err)
+	}
+
+	addr, err := h.Network().InterfaceListenAddresses()
```

Comment:
I'm finding it confusing to read as `addr` everywhere when it's an array of addresses. Can it be pluralised to `addrs` as it was previously? 🙏

Note: same with the accessor method `Addr()`.

Reply:
I changed it. But I'll say that the Go style guide forbids naming variables in the plural (even for slices).

Reply:
WAT! no way!

Reply:
I couldn't find a reference for this. So you can ignore it. The original Go authors' code did not use plurals ... it looks like this has changed.